# EXERCISE & TASK 4.5
# Data Consistency Checks

#### Table of Content

1. Importing Libraries 
2. Dataframes
3. Exercise 4.5
4. Task 4.5
5. Exporting Dataframes

## 1. Importing Libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

## 2. Dataframes

In [2]:
path = r'/Users/charliegehant/Dropbox/4. Career Foundry/Achievement 4 - Python Fundamentals for Data Analysts/02-2022 Instacart Basket Analysis'

In [3]:
path

'/Users/charliegehant/Dropbox/4. Career Foundry/Achievement 4 - Python Fundamentals for Data Analysts/02-2022 Instacart Basket Analysis'

In [4]:
# dataframe products file original

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# dataframe orders file prepared/wrangled

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## 3. Exercise 4.5

In [6]:
# Create a dataframe

df_test = pd.DataFrame()

In [7]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Function for checking whether a dataframe contains any mixed-type columns

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)


mix


In [10]:
# Finding missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Creating a new DF with only the missing values rows.

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
# DF with the 16 rows missing a product name

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
# Checking what a product name is supposed to be like

df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [14]:
# Ways you can replace missing values with mean or median:
## df['column with missings'].fillna(mean value, inplace=True)
## df['column with missings'].fillna(median value, inplace=True)

In [15]:
df_prods.shape

(49693, 5)

In [16]:
# Excluding NaN / Missing values from the DF

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
# Other ways to drop missing values from a DF
## df_prods.dropna(inplace = True) >>>> this will drop all missing values form all columns.
## df_prods.dropna(subset = [‘product_name’, inplace = True) >>>> this will drop missing values from the specified column(s)
# In both cases, rather than creating an entirely new dataframe, you’re overwriting df_prods with a new version of df_prods that doesn’t contain the missing values. This is done by way of the inplace = True function

In [19]:
# The following command will look for full duplicates within your dataframe:

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
# new dataframe that doesn’t include the duplicates you just identified using the drop_duplicates() function

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

## 4. Task 4.5

> ##### D2: Run the df.describe() function on your df_prods dataframe.

> <i> Even though the question calls to check the 'df_prods' DF, I feel like I should look at 'df_prods_clean_no_dups' rather than 'df_drops', seeing as I just performed cleaning tasks to this DF. 

In [24]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### TO BE INVESTIGATED FURHTER: 
#### A. In the 'prices' column, the max is 99999, which seems like a lot for any item to be so expensive. Also the mean and the 50 and 75 percentiles are much lower figures, telling me 99999 is a abnormal outlier.
#### B. In the 'product_id' column, the max (49688) is greater than the count of products (49672). Whilst the difference is not big and this could be explained by old products having been taken off sales, this is something I would investigate and/or bring up to my superior/stakeholders.

In [25]:
# Investigating the problem with the price column

df_investigate = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999]

In [26]:
df_investigate.shape

(1, 5)

In [27]:
# Finding the culprit! $99999 for some milk is quite the expense. 

df_investigate

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

> ##### D3: Check for mixed-type data in your df_ords dataframe.

In [28]:
# Checking for mixed type data

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

> ##### D4: If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

> <i> >>>> The function is returning no data / columns, we can therefore assume there are no mixed-type data  in the df_ords dataframe

> ##### D5: Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [29]:
# Checking for null values

df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [30]:
df_ords.shape

(3421083, 7)

> <i> >>>> There are 206209 missing values in the 'days_since_prior_order' column. I assume this is because all these orders were the user's (client's) first order, therefor therewas no prior order. 

In [31]:
# Creating a new DF with the missing values

df_nan_q5 = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [32]:
df_nan_q5

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
11        2168274        2    prior             1                   2   
26        1374495        3    prior             1                   1   
39        3343014        4    prior             1                   6   
45        2717275        5    prior             1                   3   
...           ...      ...      ...           ...                 ...   
3420930    969311   206205    prior             1                   4   
3420934   3189322   206206    prior             1                   3   
3421002   2166133   206207    prior             1                   6   
3421019   2227043   206208    prior             1                   1   
3421069   3154581   206209    prior             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

> <i> >>>> Here I have confirmed my theory, as all the missing values from the 'days_since_prior_order' column are linked with an 'order_number' = 1, which means it is that user's first ever order. 

> ##### D6: Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

> <i> >>>> CORRECTION REQUEST from TUTOR; replace NaN with 0.

In [33]:
# Creating a new DF to replace NaN values to 0

df_ords_clean = df_ords

In [34]:
df_ords_clean.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [35]:
# Filling in missing values (NaN) with 0

df_ords_clean['days_since_prior_order'].fillna(0, inplace=True)

In [36]:
df_ords_clean.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     0.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

> ##### D7: Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [37]:
# Finding duplicates

df_dups_ords = df_ords[df_ords.duplicated()]

In [38]:
# There are no duplicates!

df_dups_ords

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

> ##### D8: Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [39]:
# I cannot address the duplicates, because there are none.

df_dups_ords.shape

(0, 7)

## 5. Exporting Dataframes

> ##### D9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [40]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'), index=False)

In [41]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index=False)